enhancements/baremetal/baremetal-poison-pill-remediation.md

```diff
@@ -0,0 +1,189 @@
+1. Resource starvation, preventing our DaemonSet from initiating the check 
+1. Authentication failure to API (ServiceAccount deleted?)
+1. RBAC disallows read operation on Machine (Role deleted or changed
+
+We only want to remediate if the problem is local (#1 and #4). Otherwise, we could create a "remediation storm" where
+all nodes are trying to remediate themselves, even if the problem is not theirs.
+
+The semantics of the watchdog device automatically handle #4 and to distinguish between the remaining three situations,
+the node will ask its peers whether they think it is healthy.
+
+The possible responses from a peer are:
+1. You're unhealthy (the annotation exists)
+1. You're healthy (the annotation does not exist)
+1. The peer also cannot contact the api-server
+1. No response (timeout or any TCP/HTTP error)
+
+To avoid saturating the network, the node will pick *min(cluster_size/10, cluster_size-1)* random nodes, and ask these
+nodes if they think it’s healthy using HTTP.
+Each pod will listen on a specific node-port, and will run an HTTP web server that will be able to tell other peers if
+the machine is healthy or not.
+If two or more peers responded with a different result, we take an optimistic approach - one result of
+*"you’re healthy"* is enough to determine that the node is healthy.
+If all randomized peers returned *"you’re unhealthy"* - the node considers itself unhealthy.
+If all randomized peers returned an error (#3 or #4), another set of random nodes will be used to determine health
+status.
+The node considers itself healthy if more than 50% of the nodes responded with #3.
+The node considers itself unhealthy if we do not receive any peer responses.
+The above algorithm to determine health state of a node is demonstrated in the following flow chart:
+
+![Machine health check](./baremetal-poison-pill-health-flow.png)
+
+The state required to run this algorithm will be stored in-memory, including nodes addresses, retries count and so on.
+
+Remediation will be done by using a watchdog device to trigger a reset of the node.
+
+In the absence of a watchdog device, including softdog, a software reboot will be issued.
+
+While reboot might not fix the health issue of the node, it will allow the cluster to assume that the workload on the failing node is no longer running, such that stateful applications scheduled on that node can be rescheduled to other healthy nodes.
+
+Steps taken by the poison pill pod (on all nodes in the cluster) for any unhealthy node:
+1. Mark the node as unschedulable (this prevents the node to run new workload after reboot)
+1. Add current time to annotation
+1. Unhealthy node reboot itself
+1. After *timeout-to-assume-node-rebooted* (either configurable or calculated by software) - delete the unhealthy node (to signal the cluster that the workload is no longer running there, and it is safe to run it elsewhere)
```

Comment:
Do we also want to pair this with a set of re-tries? Or a buffer +/- 20%?

Reply:
hmm can you elaborate? retries for which operation?